File: NX1899_NX1903_NX1907_NX1911_NX1915_NX1919/UGOPEN/SampleNXOpenApplications/.NET/Routing/Routing_Example_DesignRules.vb
'===================================================================================================
'               Copyright 2017 Siemens Product Lifecycle Management Software Inc.
'===================================================================================================
'
'    NOTE:  NX Development provides programming examples for illustration only.
'    NX Development assumes you are familiar with the programming language
'    demonstrated in these examples, and the tools used to create and debug NX/Open
'    programs. GTAC support professionals can help explain the functionality of
'    a particular procedure, but neither GTAC nor NX Development will modify
'    these examples to provide added functionality or construction procedures.
'
'File description:
'
'    A Visual Basic module that implements example Routing Design Rules.
'
'
'===================================================================================================
Option Strict On
Imports System
Imports NXOpen
Imports NXOpen.Routing
Imports NXOpen.Routing.Electrical
Imports NXOpen.Routing.CustomManager
Imports NXOpen.Preferences

Module DesignRules

    Dim theSession As Session = Nothing
    Dim listingWindow As ListingWindow = Nothing
    Dim customManager As CustomManager = Nothing
    Dim createPathRuleId As Integer = 0
    Dim subdivideSegmentRuleId As Integer = 0

    '-----------------------------------------------------------------------------------------------
    ' Called when NX starts up to register the Design Rules.
    Function Startup(ByVal args As String()) As Integer

        If theSession Is Nothing Then
            theSession = Session.GetSession()
        End If

        If listingWindow Is Nothing Then
            listingWindow = theSession.ListingWindow
        End If

        If customManager Is Nothing Then
            customManager = theSession.RouteCustomManager()
        End If

        RegisterDesignRules()

        Startup = 0
    End Function

    '------------------------------------------------------------------------------------------
    ' Called from File->Execute->NX Open. Useful for testing.
    Sub Main()

        If theSession Is Nothing Then
            theSession = Session.GetSession()
        End If

        If listingWindow Is Nothing Then
            listingWindow = theSession.ListingWindow
        End If

        If customManager Is Nothing Then
            customManager = theSession.RouteCustomManager()
        End If

        RegisterDesignRules()
    End Sub

    '------------------------------------------------------------------------------------------
    ' Called by both Main And Startup to register the Design Rules.
    Sub RegisterDesignRules()

        Dim createPathName As String = "Create Path"
        Dim subdivideSegmentName As String = "Subdivide Segment"
        Dim createPathDescription As String = "The Create Path design rule makes sure each segment on the path is not too long."
        Dim subdivideSegmentDescription As String = "The Subdivide Segment design rule makes sure each new segment is not too short."

        createPathRuleId = customManager.AddDesignRule(customManager.DesignRuleReason.CreatePath, createPathName, createPathDescription, AddressOf CreatePathDesignRule)
        createPathRuleId = customManager.AddDesignRule(customManager.DesignRuleReason.SubdivideSegment, subdivideSegmentName, subdivideSegmentDescription, AddressOf SubdivideSegmentDesignRule)

    End Sub

    '-----------------------------------------------------------------------------------------------
    Sub CreatePathDesignRule(ByVal reason As DesignRuleReason, ByVal nxObjects As NXObject())
        Dim createPathName As String = "Create Path"

        Dim maximumLength As Double = 1000.0

        If reason = DesignRuleReason.Interactive Then

            ' For interactive checks, check all the segments in the work part.
            Dim segmentManager As SegmentManager = theSession.Parts.Work.SegmentManager
            Dim segmentCollection As ISegmentCollection = segmentManager.Segments

            For Each segment As ISegment In segmentCollection
                CheckForLongSegments(createPathName, maximumLength, segment)
            Next

            Return
        End If

        ' Otherwise, check the segments in the given list of objects.
        For Each nxObject As NXObject In nxObjects
            If TypeOf nxObject Is SplineSegment Then
                CheckForLongSegments(createPathName, maximumLength, CType(nxObject, ISegment))
            End If

        Next

    End Sub

    '-----------------------------------------------------------------------------------------------
    ' Checks the length of the given segment against the maximum allowed length.
    '
    ' Creates a violation for the given reason if the segment Is too long.
    '
    ' \param[in]
    '      maximumLength
    '          The longest length allowed.
    '
    ' \param[in]
    '      segment
    '          The segment to check.
    Sub CheckForLongSegments(ByVal designRuleName As String, ByVal maximumLength As Double, ByVal segment As ISegment)

        If segment.Length < maximumLength Then
            Return
        End If

        Dim objects() As Curve = New Curve() {CType(segment, Curve)}

        customManager.CreateViolationForReason(designRuleName,
                                               customManager.DesignRuleReason.CreatePath,
                                               "This path is too long.",
                                               "Long paths need to be broken up and supported by clamps.",
                                               objects)
    End Sub

    '-----------------------------------------------------------------------------------------------
    Sub SubdivideSegmentDesignRule(ByVal reason As DesignRuleReason, ByVal nxObjects As NXObject())
        Dim subdivideSegmentName As String = "Subdivide Segment"

        Dim userPreferences As RoutingUserPreferences = theSession.Preferences.RoutingApplicationView.RoutingUserPreferences

        Dim minimumLength As Double = 0.0
        If Not userPreferences.GetDoublePreference("ROUTE_ELEC_MIN_STRAIGHT_LENGTH_COEFF", minimumLength) Then
            minimumLength = 10.0
        End If

        If reason = DesignRuleReason.Interactive Then

            ' For interactive checks, check all the segments in the work part.
            Dim segmentManager As SegmentManager = theSession.Parts.Work.SegmentManager
            Dim segmentCollection As ISegmentCollection = segmentManager.Segments

            For Each segment As ISegment In segmentCollection
                CheckForShortSegments(subdivideSegmentName, minimumLength, segment)
            Next

            Return

        End If

        ' Otherwise, check the segments in the given list of objects.
        For Each nxObject As NXObject In nxObjects
            If TypeOf nxObject Is SplineSegment Then
                CheckForShortSegments(subdivideSegmentName, minimumLength, CType(nxObject, ISegment))
            End If

        Next

    End Sub

    '-----------------------------------------------------------------------------------------------
    ' Checks the length of the given segment against the minimum allowed length.
    '
    ' Creates a violation for the given reason if the segment Is too short.
    '
    ' \param[in]
    '      maximumLength
    '          The shortest length allowed.
    '
    ' \param[in]
    '      segment
    '          The segment to check.
    Sub CheckForShortSegments(ByVal designRuleName As String, ByVal maximumLength As Double, ByVal segment As ISegment)

        If segment.Length > maximumLength Then
            Return
        End If

        Dim objects() As Curve = New Curve() {CType(segment, Curve)}

        customManager.CreateViolationForReason(designRuleName,
                                               customManager.DesignRuleReason.SubdivideSegment,
                                               "This segment is too short.",
                                               "Short segments violate our design standards.",
                                               objects)
    End Sub

    '-----------------------------------------------------------------------------------------------
    ' Tells NX when to unload your application.
    ' For Routing callbacks, plugins, And Design Rules, this MUST return AtTermination.
    Public Function GetUnloadOption(ByVal dummy As String) As Integer
        GetUnloadOption = NXOpen.Session.LibraryUnloadOption.AtTermination
    End Function

End Module

File: NX12.0.2.9/UGOPEN/SampleNXOpenApplications/.NET/Routing/Routing_Example_DesignRules.cs
/*==================================================================================================
               Copyright 2017 Siemens Product Lifecycle Management Software Inc.
====================================================================================================

    NOTE:  NX Development provides programming examples for illustration only.
    NX Development assumes you are familiar with the programming language
    demonstrated in these examples, and the tools used to create and debug NX/Open
    programs. GTAC support professionals can help explain the functionality of
    a particular procedure, but neither GTAC nor NX Development will modify
    these examples to provide added functionality or construction procedures.

File description:

    A C# class that implements example Routing Design Rules.


==================================================================================================*/
using System;
using NXOpen;
using NXOpen.Preferences;
using NXOpen.Routing;
using NXOpen.Routing.Electrical;
using System.Collections.Generic;

namespace Routing
{
    public class DesignRules
    {
        // Static class members
        static Session theSession = null;
        static ListingWindow listingWindow = null;
        static CustomManager customManager = null;
        static int createPathRuleId = 0;
        static int subdivideSegmentRuleId = 0;

        //------------------------------------------------------------------------------------------
        // Called when NX starts up to register the Design Rules.
        public static int Startup()
        {
            if (theSession == null)
                theSession = Session.GetSession();

            if (listingWindow == null)
                listingWindow = theSession.ListingWindow;

            if (customManager == null)
                customManager = theSession.RouteCustomManager;

            RegisterDesignRules();

            return 0;
        }

        //------------------------------------------------------------------------------------------
        // Called from File->Execute->NX Open. Useful for testing.
        public static void Main()
        {
            if (theSession == null)
                theSession = Session.GetSession();

            if (listingWindow == null)
                listingWindow = theSession.ListingWindow;

            if (customManager == null)
                customManager = theSession.RouteCustomManager;

            RegisterDesignRules();
        }

        //------------------------------------------------------------------------------------------
        // Called by both Main and Startup to register the Design Rules.
        private static void RegisterDesignRules()
        {
            string createPathName = "Create Path";
            string subdivideSegmentName = "Subdivide Segment";
            string createPathDescription = "The Create Path design rule makes sure each segment on the path is not too long.";
            string subdivideSegmentDescription = "The Subdivide Segment design rule makes sure each new segment is not too short.";

            createPathRuleId = customManager.AddDesignRule(CustomManager.DesignRuleReason.CreatePath, createPathName, createPathDescription, CreatePathDesignRule);
            subdivideSegmentRuleId = customManager.AddDesignRule(CustomManager.DesignRuleReason.SubdivideSegment, subdivideSegmentName, subdivideSegmentDescription, SubdivideSegmentDesignRule);
        }

        //------------------------------------------------------------------------------------------
        static void CreatePathDesignRule
        (
            CustomManager.DesignRuleReason reason,
            NXObject[] nxObjects
        )
        {
            string createPathName = "Create Path";

            double maximumLength = 1000.0;

            if (reason == CustomManager.DesignRuleReason.Interactive)
            {
                // For interactive checks, check all the segments in the work part.
                SegmentManager segmentManager = theSession.Parts.Work.SegmentManager;
                ISegmentCollection segmentCollection = segmentManager.Segments;

                foreach (ISegment segment in segmentCollection)
                    checkForLongSegments(createPathName, maximumLength, segment);

                return;
            }

            // Otherwise, check the segments in the given list of objects.
            foreach (NXObject nxObject in nxObjects)
            {
                if (nxObject.GetType() == typeof(SplineSegment))
                    checkForLongSegments(createPathName, maximumLength, (ISegment)nxObject);
            }
        }

        //------------------------------------------------------------------------------------------
        // Checks the length of the given segment against the maximum allowed length.
        //
        // Creates a violation for the given reason if the segment is too long.
        //
        // \param[in]
        //      maximumLength
        //          The longest length allowed.
        //
        // \param[in]
        //      segment
        //          The segment to check.
        static void checkForLongSegments
        (
            string designRuleName,
            double maximumLength,
            ISegment segment
        )
        {
            if (segment.Length < maximumLength)
                return;

            List<NXObject> objects = new List<NXObject>();
            objects.Add((Curve)segment);

            customManager.CreateViolationForReason(designRuleName,
                                                   CustomManager.DesignRuleReason.CreatePath,
                                                   "This path is too long.",
                                                   "Long paths need to be broken up and supported by clamps.",
                                                   objects.ToArray());
        }

        //------------------------------------------------------------------------------------------
        static void SubdivideSegmentDesignRule
        (
            CustomManager.DesignRuleReason reason,
            NXObject[] nxObjects
        )
        {
            string subdivideSegmentName = "Subdivide Segment";

            RoutingUserPreferences userPreferences = theSession.Preferences.RoutingApplicationView.RoutingUserPreferences;

            double minimumLength = 0.0;
            if (!userPreferences.GetDoublePreference("ROUTE_ELEC_MIN_STRAIGHT_LENGTH_COEFF", out minimumLength))
                minimumLength = 10.0;

            if (reason == CustomManager.DesignRuleReason.Interactive)
            {
                // For interactive checks, check all the segments in the work part.
                SegmentManager segmentManager = theSession.Parts.Work.SegmentManager;
                ISegmentCollection segmentCollection = segmentManager.Segments;

                foreach (ISegment segment in segmentCollection)
                    checkForShortSegments(subdivideSegmentName, minimumLength, segment);

                return;
            }

            // Otherwise, check the segments in the given list of objects.
            foreach (NXObject nxObject in nxObjects)
            {
                if (nxObject.GetType() == typeof(SplineSegment))
                    checkForShortSegments(subdivideSegmentName, minimumLength, (ISegment)nxObject);
            }
        }

        //------------------------------------------------------------------------------------------
        // Checks the length of the given segment against the minimum allowed length.
        //
        // Creates a violation for the given reason if the segment is too short.
        //
        // \param[in]
        //      minimumLength
        //          The shortest length allowed.
        //
        // \param[in]
        //      segment
        //          The segment to check.
        static void checkForShortSegments
        (
            string designRuleName,
            double minimumLength,
            ISegment segment
        )
        {
            if (segment.Length > minimumLength)
                return;

            List<NXObject> objects = new List<NXObject>();
            objects.Add((Curve)segment);

            customManager.CreateViolationForReason(designRuleName,
                                                   CustomManager.DesignRuleReason.SubdivideSegment,
                                                   "This segment is too short.",
                                                   "Short segments violate our design standards.",
                                                   objects.ToArray());
        }

        //------------------------------------------------------------------------------------------
        // Tells NX when to unload your application.
        // For Routing callbacks, plugins, and Design Rules, this MUST return AtTermination.
        public static int GetUnloadOption(string arg)
        {
            return System.Convert.ToInt32(Session.LibraryUnloadOption.AtTermination);
        }
    }
}
